c-scape/src/mem/string.rs

```diff
@@ -0,0 +1,322 @@
+    libc!(libc::strlen(s));
+
+    let mut w = s;
+    while *w != NUL {
+        w = w.add(1);
+    }
+
+    w.offset_from(s) as usize
+}
+
+#[no_mangle]
+unsafe extern "C" fn strncat(d: *mut c_char, mut s: *const c_char, mut n: usize) -> *mut c_char {
+    libc!(libc::strcat(d, s));
+
+    let mut w = strchr(d, 0);
+
+    while n > 0 && *s != NUL {
+        n = n - 1;
+
+        *w = *s;
+
+        w = w.add(1);
+        s = s.add(1);
+    }
+    *w = 0;
+
+    d
+}
+
+#[no_mangle]
+unsafe extern "C" fn strncmp(mut s1: *const c_char, mut s2: *const c_char, mut n: usize) -> c_int {
+    libc!(libc::strcmp(s1, s2));
+
+    while n > 0 && *s1 != NUL && *s2 != NUL {
+        n = n - 1;
+
+        if *s1 != *s2 {
+            break;
+        }
+
+        s1 = s1.add(1);
+        s2 = s2.add(1);
+    }
+
+    (*s1 - *s2) as c_int
+}
+
+#[no_mangle]
+unsafe extern "C" fn strndup(s: *const c_char, n: usize) -> *mut c_char {
+    libc!(libc::strndup(s, n));
+
+    let len = strnlen(s, n);
+    let d = malloc(len + 1);
+    if !d.is_null() {
+        memcpy(d, s.cast(), len);
+    }
+
+    let ret = d.cast::<c_char>();
+    *ret.add(len) = 0;
+    ret
+}
+
+#[no_mangle]
+unsafe extern "C" fn strnlen(s: *const c_char, mut n: usize) -> usize {
+    libc!(libc::strlen(s));
+
+    let mut w = s;
+    while n > 0 && *w != NUL {
+        n = n - 1;
+        w = w.add(1);
+    }
+
+    w.offset_from(s) as usize
+}
+
+#[no_mangle]
+unsafe extern "C" fn strpbrk(s: *const c_char, m: *const c_char) -> *mut c_char {
+    libc!(libc::strpbrk(s, m));
+
+    let s = s.add(strcspn(s, m)) as *mut _;
+
+    if *s != NUL {
+        return ptr::null_mut();
+    }
+
+    s
+}
+
+#[no_mangle]
+unsafe extern "C" fn strrchr(s: *const c_char, c: c_int) -> *mut c_char {
+    libc!(libc::strrchr(s, c));
+
+    let mut s = s as *mut c_char;
+    let mut ret = ptr::null_mut::<c_char>();
+    loop {
+        s = strchr(s, c);
+        if s.is_null() {
+            break;
+        }
+
+        ret = s;
+        s = s.add(1);
+    }
+
+    ret
+}
+
+#[no_mangle]
+unsafe extern "C" fn strspn(s: *const c_char, m: *const c_char) -> usize {
+    libc!(libc::strspn(s, m));
+
+    let mut w = s;
+    while *w != NUL {
+        let mut m = m;
+        while *m != NUL {
+            if *w == *m {
+                break;
+            }
+            m = m.add(1);
+        }
+
+        if *m == NUL {
+            break;
+        }
+
+        w = w.add(1);
+    }
+
+    w.offset_from(s) as usize
+}
+
+#[no_mangle]
+unsafe extern "C" fn strtok(s: *mut c_char, m: *const c_char) -> *mut c_char {
+    libc!(libc::strtok(s, m));
+
+    #[repr(transparent)]
+    #[derive(Copy, Clone)]
+    struct SyncCChar(*mut c_char);
+    unsafe impl Sync for SyncCChar {}
+
+    static STORAGE: SyncUnsafeCell<SyncCChar> = SyncUnsafeCell::new(SyncCChar(ptr::null_mut()));
+
+    strtok_r(
+        s,
+        m,
+        SyncUnsafeCell::raw_get(ptr::addr_of!(STORAGE)) as *mut *mut c_char,
```

Comment:
`&STORAGE` is fine here, right?

Reply:
To be honest I'm not sure. I could also use `static mut` here but all the warnings I've read have dissuaded me, so I erred on the side of caution.

Reply:
I think it should be fine due to the reference being to a `SyncUnsafeCell` and not the inner value.